zbus: assert when inside an ISR and time out is not zero

brandon-exact opened the pull request:
Currently various zbus functions silently change the timeout
to zero  when inside an ISR. If a developer is not aware
of this it could lead to unexpected behaviour or a
publish/read/notify failing.

Also in the zbus docs it states to only use a timeout of
zero when inside a ISR multiple times.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -380,10 +380,8 @@ int zbus_chan_pub(const struct zbus_channel *chan, const void *msg, k_timeout_t
 
 	_ZBUS_ASSERT(chan != NULL, "chan is required");
 	_ZBUS_ASSERT(msg != NULL, "msg is required");
-
-	if (k_is_in_isr()) {
-		timeout = K_NO_WAIT;
-	}
```
Please do not remove it. We use semaphore internally, and a semaphore cannot receive a timeout different from `K_NO_WAIT` inside ISRs. So, the assert is very welcome, but removing the `if` would make the code less robust.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -416,10 +414,8 @@ int zbus_chan_read(const struct zbus_channel *chan, void *msg, k_timeout_t timeo
 {
 	_ZBUS_ASSERT(chan != NULL, "chan is required");
 	_ZBUS_ASSERT(msg != NULL, "msg is required");
-
-	if (k_is_in_isr()) {
-		timeout = K_NO_WAIT;
-	}
```
Ditto.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -438,10 +434,8 @@ int zbus_chan_notify(const struct zbus_channel *chan, k_timeout_t timeout)
 	int err;
 
 	_ZBUS_ASSERT(chan != NULL, "chan is required");
-
-	if (k_is_in_isr()) {
-		timeout = K_NO_WAIT;
-	}
```
Ditto.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -462,10 +456,8 @@ int zbus_chan_notify(const struct zbus_channel *chan, k_timeout_t timeout)
 int zbus_chan_claim(const struct zbus_channel *chan, k_timeout_t timeout)
 {
 	_ZBUS_ASSERT(chan != NULL, "chan is required");
-
-	if (k_is_in_isr()) {
-		timeout = K_NO_WAIT;
-	}
```
Ditto.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -380,10 +380,8 @@ int zbus_chan_pub(const struct zbus_channel *chan, const void *msg, k_timeout_t
-	if (k_is_in_isr()) {
-		timeout = K_NO_WAIT;
-	}
+	_ZBUS_ASSERT(!(k_is_in_isr() && !K_TIMEOUT_EQ(timeout, K_NO_WAIT)),
```
I am not a big fan of ternary, but in this case, where the readability of the expression is not clear without it, I would rather suggest using something like: 
```c
_ZBUS_ASSERT(k_is_in_isr() ? K_TIMEOUT_EQ(timeout, K_NO_WAIT) : true);
```

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -416,10 +414,8 @@ int zbus_chan_read(const struct zbus_channel *chan, void *msg, k_timeout_t timeo
-	if (k_is_in_isr()) {
-		timeout = K_NO_WAIT;
-	}
+	_ZBUS_ASSERT(!(k_is_in_isr() && !K_TIMEOUT_EQ(timeout, K_NO_WAIT)),
+		     "Timeout must be K_NO_WAIT inside an ISR");
```
Ditto.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -438,10 +434,8 @@ int zbus_chan_notify(const struct zbus_channel *chan, k_timeout_t timeout)
-	if (k_is_in_isr()) {
-		timeout = K_NO_WAIT;
-	}
+	_ZBUS_ASSERT(!(k_is_in_isr() && !K_TIMEOUT_EQ(timeout, K_NO_WAIT)),
+		     "Timeout must be K_NO_WAIT inside an ISR");
```
Ditto.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -462,10 +456,8 @@ int zbus_chan_notify(const struct zbus_channel *chan, k_timeout_t timeout)
-	if (k_is_in_isr()) {
-		timeout = K_NO_WAIT;
-	}
+	_ZBUS_ASSERT(!(k_is_in_isr() && !K_TIMEOUT_EQ(timeout, K_NO_WAIT)),
+		     "Timeout must be K_NO_WAIT inside an ISR");
```
Ditto.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -380,10 +380,8 @@ int zbus_chan_pub(const struct zbus_channel *chan, const void *msg, k_timeout_t
-		timeout = K_NO_WAIT;
-	}
+	_ZBUS_ASSERT(!(k_is_in_isr() && !K_TIMEOUT_EQ(timeout, K_NO_WAIT)),
+		     "Timeout must be K_NO_WAIT inside an ISR");
```
Just to keep consistency with the other asserts, please change the text. 
```suggestion
		     "timeout must be K_NO_WAIT inside an ISR");
```

brandon-exact committed:
zbus: assert when inside an ISR and time out is not zero

Currently various zbus functions silently change the timeout
to zero  when inside an ISR. If a developer is not aware
of this it could lead to unexpected behaviour or a
publish/read failing.

Also in the zbus docs it states to only use a timeout of
zero when inside a ISR multiple times.

Signed-off-by: Brandon Allen <brandon.allen@exacttechnology.com>

rodrigopex commented:
@brandon-exact , thanks for the contribution  🎉  🎉  🎉! I am glad to see new faces helping out zbus and Zephyr.
